onnxruntime/core/providers/vitisai/imp/ep_context_utils.cc

```diff
@@ -0,0 +1,831 @@
+  LOGS_DEFAULT(VERBOSE) << "Opset import size of original graph: " << graph_viewer.DomainToVersionMap().size();
+  // XXX: Will Steps 1/2/3 suffice for restoring a model/graph later?
+  // Any information loss or mismatch?
+  // Step 1
+  const Graph& orig_graph = graph_viewer.GetGraph();
+  // Step 2
+  const Model& orig_model = orig_graph.GetModel();
+  // Step 3
+  auto p_orig_model_proto = const_cast<Model&>(orig_model).ToProto();
+  if (p_orig_model_proto->opset_import_size() == 0) {
+    LOGS_DEFAULT(VERBOSE) << "Adding op domain version mapping: " << graph_viewer.DomainToVersionMap().size();
+    for (const auto& it : graph_viewer.DomainToVersionMap()) {
+      auto* p_opset_import = p_orig_model_proto->add_opset_import();
+      *(p_opset_import->mutable_domain()) = it.first;
+      p_opset_import->set_version(it.second);
+    }
+  }
+
+  nlohmann::json j_obj;
+  if (p_orig_model_proto->opset_import_size() > 0) {
+    LOGS_DEFAULT(VERBOSE) << "Adding op domain version mapping to JSON";
+    for (int i = 0, n = p_orig_model_proto->opset_import_size(); i < n; ++i) {
+      auto& op_set_id_proto = const_cast<ONNX_NAMESPACE::OperatorSetIdProto&>(p_orig_model_proto->opset_import(i));
+      j_obj[*op_set_id_proto.mutable_domain()] = std::to_string(op_set_id_proto.version());
+      LOGS_DEFAULT(VERBOSE) << "Added domain: " << *op_set_id_proto.mutable_domain() << ", version: " << op_set_id_proto.version();
+    }
+  }
+  j_obj["orig_graph_name"] = graph_viewer.Name();
+  j_obj["orig_model_path"] = PathToUTF8String(graph_viewer.ModelPath().ToPathString());
+
+  // XXX: `ModelProto::SerializeToString` will lose some info,
+  // e.g., ModelProto.opset_import.
+  std::string ser_buf;
+  p_orig_model_proto->SerializeToString(ser_buf);
+  LOGS_DEFAULT(VERBOSE) << "Done serializing original model";
+  j_obj["orig_model_proto_ser_str"] = ser_buf;
+
+  LOGS_DEFAULT(VERBOSE) << "Model proto JSON dumping";
+  return j_obj.dump(-1, ' ', false, nlohmann::json::error_handler_t::replace);
+}
+
+// Ref.: `CreateEpContextModel()` in the file "graph_partitioner.cc".
+ONNX_NAMESPACE::ModelProto* CreateEPContexModel(
+    const GraphViewer& graph_viewer,
+    const std::string& serialized_ctx_cache,
+    const std::string& ctx_cache_file_loc,
+    const int64_t embed_mode,
+    const std::string& backend_cache_dir,
+    const std::string& backend_cache_key,
+    bool saving_orig_graph,
+    const logging::Logger* p_logger) {
+  LOGS_DEFAULT(VERBOSE) << "[VitisAI EP]Creating EP context node";
+  // Create a new graph/model, reusing the graph name,
+  // the op-domain-to-opset-version map,
+  // and the op schema registry of the current graph.
+  // XXX: This approach (immediately below) has a memory fault issue (std::bad_alloc).
+  // auto& ep_ctx_graph = graph_viewer.CreateModel(*p_logger)->MainGraph();
+  // This apporach (immediately below) has no memory falut issue.
+  auto p_temp_model = graph_viewer.CreateModel(*p_logger);
```

Comment:
The ownership and life cycle of the p_temp_model, so as the ModelProto* returned from this method is not clear, could cause problem.
I suppose the model should be created and maintained in the EP. That's what we did in QNN EP.

Reply:
Seems like that's not the case.

The `p_ep_ctx_model_proto` (L313) has its own lifespan and does NOT rely on `p_temp_model` (L230).
Details on this conclusion above:
1) https://github.com/microsoft/onnxruntime/blob/4743803944d6f6333bf93b89a62d8083d4466710/onnxruntime/core/providers/shared_library/provider_wrappedtypes.h#L842
2) https://github.com/microsoft/onnxruntime/blob/4743803944d6f6333bf93b89a62d8083d4466710/onnxruntime/core/session/provider_bridge_ort.cc#L1041
3) https://github.com/microsoft/onnxruntime/blob/4743803944d6f6333bf93b89a62d8083d4466710/onnxruntime/core/graph/model.cc#L369-L379

---

Yes, I took a close look at how `CreateEPContextNodes()` is implemented and invoked in QNN EP and how `CreateCtxModel()` is implemented and invoked in TRT EP and **partially** did similar things here.

Reply:
A related documentation "Complexities of ownerships and lifespan of `Model`, `Graph`, and `GraphViewer`" I made before - still partially done, not completed yet.

# Pre-knowledge
## `onnxruntime` entities
- `onnxruntime::Model`
  - `std::unique_ptr<Graph> graph_`
  - `ONNX_NAMESPACE::ModelProto model_proto_`
  - Shared-lib wrappers
    - `Graph& MainGraph()`
    - `std::unique_ptr<ONNX_NAMESPACE::ModelProto> ToProto()`
- `onnxruntime::Graph`
  - `const Model& owning_model_`
  - `ONNX_NAMESPACE::GraphProto* graph_proto_`
  - Shared-lib wrappers
    - `std::unique_ptr<GraphViewer> CreateGraphViewer()`
    - `std::unique_ptr<ONNX_NAMESPACE::GraphProto> ToGraphProto()`
    - `const Model& GetModel()`
- `onnxruntime::GraphViewer`
  - `const Graph* graph_`
  - Shared-lib wrappers
    - `std::unique_ptr<Model> CreateModel()`
    - `const Graph& GetGraph()`
    - `void ToProto()`
## Protos of entities (shared-lib wrappers)
- `ONNX_NAMESPACE::ModelProto`
  - `const GraphProto& graph()`
  - `GraphProto* mutable_graph()`
- `ONNX_NAMESPACE::GraphProto`
## `onnx` entities